drivers/iio/light/cm36283.c

```diff
@@ -0,0 +1,852 @@
+	.write_event_config =	cm36283_write_prox_event_config,
+};
+
+static int cm36283_setup(struct cm36283_data *chip)
+{
+	struct i2c_client *client = chip->client;
+	struct iio_dev *indio_dev = i2c_get_clientdata(client);
+	int ret;
+
+	ret = i2c_smbus_read_byte_data(client, CM36283_REG_ID_FLAG);
+	if (ret < 0)
+		return dev_err_probe(&client->dev, ret,
+				     "Failed to read ID flag register\n");
+
+	switch (ret) {
+	case CM36283:
+		chip->model = CM36283;
+		indio_dev->name = "cm36283";
+		chip->ps_it_times = cm36283_ps_it_times;
+		chip->num_ps_it = ARRAY_SIZE(cm36283_ps_it_times);
+		chip->max_ps_value = CM36283_MAX_PS_VALUE;
+		break;
+	case CM36686:
+		chip->model = CM36686;
+		indio_dev->name = "cm36686";
+		chip->ps_it_times = cm36686_ps_it_times;
+		chip->num_ps_it = ARRAY_SIZE(cm36686_ps_it_times);
+		chip->max_ps_value = CM36686_MAX_PS_VALUE;
+		break;
+	default:
+		dev_err(&client->dev, "Unknown sensor\n");
+		return -ENODEV;
+	}
+
+	dev_info(&client->dev, "Detected sensor model: %s\n", indio_dev->name);
+
+	chip->als_conf = CM36283_ALS_ENABLE;
+	ret = i2c_smbus_write_word_data(client, CM36283_REG_ALS_CONF,
+					chip->als_conf);
+	if (ret < 0)
+		return dev_err_probe(&client->dev, ret,
+				     "Failed to enable ambient light sensor: %d\n", ret);
+
+	switch (chip->model) {
+	case CM36283:
+		chip->ps_conf[CM36283_PS_CONF1] =
+			CM36283_PS_INT_IN | CM36283_PS_INT_OUT |
+			CM36283_PS_DR_1_320 | CM36283_PS_IT_1_3T |
+			CM36283_PS_PERS_2;
+		break;
+	case CM36686:
+		chip->ps_conf[CM36283_PS_CONF1] =
+			CM36686_PS_INT_IN | CM36686_PS_INT_OUT |
+			CM36686_PS_DR_1_320 | CM36686_PS_IT_2_5T |
+			CM36686_PS_PERS_2;
+		break;
+	}
+
+	ret = i2c_smbus_write_word_data(client, CM36283_REG_PS_CONF1,
+					chip->ps_conf[CM36283_PS_CONF1]);
+	if (ret < 0)
+		return dev_err_probe(&client->dev, ret,
+				     "Failed to enable proximity sensor: %d\n", ret);
+
+	switch (chip->model) {
+	case CM36283:
+		chip->ps_conf[CM36283_PS_CONF3] =
+			CM36283_PS_SMART_PERS_ENABLE;
+		break;
+	case CM36686:
+		chip->ps_conf[CM36283_PS_CONF3] =
+			CM36686_PS_SMART_PERS_ENABLE |
+			CM36686_LED_I_100;
+		break;
+	}
+
+	ret = i2c_smbus_write_word_data(client, CM36283_REG_PS_CONF3,
+					chip->ps_conf[CM36283_PS_CONF3]);
+	if (ret < 0)
+		return dev_err_probe(&client->dev, ret,
+				     "Failed to enable PS: %d\n", ret);
+
+	ret = device_property_read_u32(&client->dev, "capella,proximity-close-threshold",
+					    &chip->ps_close);
+
+	if (ret < 0)
+		return dev_err_probe(&client->dev, ret,
+		       "Failed to read close proximity threshold: %d\n", ret);
+
+	ret = device_property_read_u32(&client->dev, "capella,proximity-away-threshold",
+					    &chip->ps_away);
+
+	if (ret < 0)
+		return dev_err_probe(&client->dev, ret,
+		       "Failed to read away proximity threshold: %d\n", ret);
+
+	ret = cm36283_set_prox_thresh(chip);
+	if (ret) {
+		dev_err(&client->dev, "Failed to set proximity thresholds: %d\n", ret);
```

Comment:
This is very much my personal preference, but I really like to use `%pe` "error pointer" formatter which expands to a textual version of the error code (i.e. if ret is -22 it will say -EINVAL)


```suggestion
		dev_err(&client->dev, "Failed to set proximity thresholds: %pe\n", ERR_PTR(ret));
```